mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/proxy/MQTTProxyProtocolMethodProcessor.java

```diff
@@ -245,6 +246,87 @@ public void processSubscribe(MqttSubscribeMessage msg) {
+                                .addSubscription(subscription.qualityOfService(), decodedTopicName)
+                                .properties(mqttProperties)
+                                .build();
+                        doSubscribe(message, false)
+                                .thenRun(() -> log.info("[{}] Subscribe new topic [{}] success",
+                                        connection.getClientId(), Codec.decode(changedTopicName.toString())))
+                                .exceptionally(ex -> {
+                                    log.error("[{}][{}] Subscribe new topic [{}] Fail, the message is {}",
+                                            connection.getClientId(), topicFilter, changedTopicName, ex.getMessage());
+                                    return null;
+                                });
+                    }
+
+                    @Override
+                    public void onTopicUnload(TopicName topicName) {
+                        // NO-OP
+                    }
+                });
+            }
+        }
+    }
+
+    private CompletableFuture<Void> doSubscribe(MqttSubscribeMessage message, boolean incrementCounter) {
+        return PulsarTopicUtils.asyncGetTopicsForSubscribeMsg(message, proxyConfig.getDefaultTenant(),
+                        proxyConfig.getDefaultNamespace(),
+                        pulsarService, proxyConfig.getDefaultTopicDomain())
+                .thenCompose(topics -> {
+                    int packetId = message.variableHeader().messageId();
+                    if (CollectionUtils.isEmpty(topics)) {
+                        SubscribeAck subscribeAck = SubscribeAck.builder()
+                                .packetId(packetId)
+                                .success(true)
+                                .grantedQoses(new ArrayList<>(message.payload().topicSubscriptions().stream()
+                                        .map(MqttTopicSubscription::qualityOfService)
+                                        .collect(Collectors.toSet())))
+                                .build();
+                        connection.getAckHandler().sendSubscribeAck(connection, subscribeAck);
+                        return CompletableFuture.completedFuture(null);
+                    }
+                    List<CompletableFuture<Void>> subscribeFutures = topics.stream()
+                            .map(topic -> {
+                                CompletableFuture<Void> future = writeToBroker(topic, message);
+                                if (incrementCounter) {
+                                    future.thenAccept(__ ->
+                                            increaseSubscribeTopicsCount(packetId, 1));
+                                }
+                                return future;
+                            }).collect(Collectors.toList());
+                    return FutureUtil.waitForAll(Collections.unmodifiableList(subscribeFutures));
```

Comment:
FutureUtil.waitForAll will return a new future, so need to wrap this by `unmodifiableList`

Reply:
fixed